File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.cpp
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        bool flag=0;
        queue<TreeNode*>q;
        
        q.push(root);
        vector<vector<int>>ans;
        if(!root) return ans;

        while(!q.empty()){
            int size=q.size();
            vector<int>temp(size);

            for(int i=0;i<size;i++){
                TreeNode* node=q.front();
                q.pop();

                int index=(flag==0) ? i:size-1-i;
                temp[index]=node->val;

                if(node->left) q.push(node->left);
                if(node->right) q.push(node->right);

            }
            flag=!flag;
            ans.push_back(temp);
        }   
        return ans;

    }
};

File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {


       List<List<Integer>> ans = new ArrayList<>();

        if(root==null) {
            return ans;
        }

    int level=0;
    Queue<TreeNode> q = new LinkedList<>();
    q.offer(root);
    while(!q.isEmpty())
    {
        int size=q.size();
        List<Integer> res = new ArrayList<>();

        for(int i=0;i<size;i++){
            root=q.poll();
            res.add(root.val);

            if(root.left!=null){
                q.offer(root.left);
            }

            if(root.right!=null){
            q.offer(root.right);
            }
        }

        if(level%2==0){
            ans.add(new ArrayList(res));
        }
        else{
            Collections.reverse(res);
            ans.add(new ArrayList(res));
        }
        level++;
    }

    return ans;
    }
}
        
    